File: pytools/examples/cpp/client/network/Packet.cpp
#include <iostream>
#include <cstring>
#include <cstdlib>
#include <cstdio>

#include <Packet.hpp>
#include <PacketUtil.hpp>


Packet::Packet()
{
	m_nCursor	= 0;

	m_len 		= 0;
	m_cap		= SLICE_SIZE;

	m_body 		= new BYTE[m_cap];
}

Packet::Packet(BYTE* buff)
{
	m_body = buff;
	m_nCursor = 0;
	m_len = 0;
}

Packet::~Packet()
{
	delete(m_body);
}


void Packet::ReAllocate(U32 size)
{
	U32 new_size = m_len + size;
	if (new_size <= m_cap) { return; }

	m_cap	= ((new_size / SLICE_SIZE) + 1) * SLICE_SIZE;

	BYTE* new_buff = new BYTE[m_cap];
	memcpy(new_buff, m_body, m_len);

	delete(m_body);

	m_body	= new_buff;
}



void Packet::Encode(U16 packetId)
{
	U32 len 		= 0;
	U32 bodyLen 	= 0;

	BYTE* body_buff = ReadBytes(&bodyLen);

	U32 packetLen 	= bodyLen + 2;
	len = packetLen + 2;

	BYTE* len_buff = PacketUtil::GetBytes((U16)packetLen);
	BYTE* packet_id_buff = PacketUtil::GetBytes((U16)packetId);

	BYTE* all_buff = new BYTE[len];

	memcpy(all_buff, len_buff, 2);
	memcpy(all_buff + 2, packet_id_buff, 2);
	memcpy(all_buff + 4, body_buff, bodyLen);

	delete(m_body);

	m_body 		= all_buff;
	m_nCursor	= len;
	m_len		= len;
	m_cap		= len;

	delete(len_buff);
	delete(packet_id_buff);
	delete(body_buff);
}


void Packet::Echo()
{
	printf("Packet body: [");
	U32 last_idx = m_len - 1;
	U8 data = 0;
	for (int i = 0; i < m_len; i++)
	{
		memcpy((void*)(&data), m_body + i, 1);
		if (i == last_idx)
		{
			printf("%d", data);
		}
		else
		{
			printf("%d,", data);
		}
	}
	printf("]\n");
}


void Packet::WriteU8(U8 data)
{  
	ReAllocate(1);

	BYTE* bytes 		= PacketUtil::GetBytes(data);
	memcpy(m_body + m_nCursor, bytes, 1);
	delete(bytes);

	m_nCursor 	+= 1;
	m_len		+= 1;
}

void Packet::WriteI8(I8 data)
{
	ReAllocate(1);

	BYTE* bytes 		= PacketUtil::GetBytes(data);
	memcpy(m_body + m_nCursor, bytes, 1);
	delete(bytes);

	m_nCursor 	+= 1;
	m_len 		+= 1;
}

void Packet::WriteU16(U16 data)
{
	ReAllocate(2);

	BYTE* bytes 		= PacketUtil::GetBytes((U16)data);
	memcpy(m_body + m_nCursor, bytes, 2);
	delete(bytes);
	
	m_nCursor	+= 2;
	m_len 		+= 2;
}

void Packet::WriteI16(I16 data)
{
	ReAllocate(2);

	BYTE* bytes 		= PacketUtil::GetBytes(data);
	memcpy(m_body + m_nCursor, bytes, 2);
	delete(bytes);
	
	m_nCursor	+= 2;
	m_len 		+= 2;
}

void Packet::WriteU32(U32 data)
{
	ReAllocate(4);

	BYTE* bytes 		= PacketUtil::GetBytes(data);
	memcpy(m_body + m_nCursor, bytes, 4);
	delete(bytes);
	
	m_nCursor	+= 4;
	m_len 		+= 4;
}

void Packet::WriteI32(I32 data)
{
	ReAllocate(4);

	BYTE* bytes 		= PacketUtil::GetBytes(data);
	memcpy(m_body + m_nCursor, bytes, 4);
	delete(bytes);
	
	m_nCursor	+= 4;
	m_len 		+= 4;
}

void Packet::WriteU64(U64 data)
{
	ReAllocate(8);

	BYTE* bytes 		= PacketUtil::GetBytes(data);
	memcpy(m_body + m_nCursor, bytes, 8);
	delete(bytes);
	
	m_nCursor	+= 8;
	m_len 		+= 8;
}

void Packet::WriteI64(I64 data)
{
	ReAllocate(8);

	BYTE* bytes 		= PacketUtil::GetBytes(data);
	memcpy(m_body + m_nCursor, bytes, 8);
	delete(bytes);
	
	m_nCursor	+= 8;
	m_len 		+= 8;
}

void Packet::WriteF32(F32 data)
{
	I32 val = *(I32*)(&data);
	WriteI32(val);
}

void Packet::WriteF64(F64 data)
{
	I64 val = *(I64*)(&data);
	WriteI64(val);
}

void Packet::WriteString(String data)
{
	U16 len 	= data.size();

	ReAllocate(len + 2);

	WriteU16(len);

	memcpy(m_body + m_nCursor, data.c_str(), len);
	m_nCursor	+= len;
	m_len 		+= len;
}

void Packet::WriteBytes(BYTE* bytes, U32 len)
{
	ReAllocate(len);

	memcpy(m_body + m_nCursor, bytes, len);

	delete(bytes);

	m_nCursor	+= len;
	m_len 		+= len;
}


U8 Packet::ReadU8()
{
	U8 data = PacketUtil::GetU8(m_body + m_nCursor);
	m_nCursor	+= 1;
	return data;
}

I8 Packet::ReadI8()
{
	I8 data = PacketUtil::GetI8(m_body + m_nCursor);
	m_nCursor	+= 1;

	return data;
}

U16 Packet::ReadU16()
{
	U16 data = PacketUtil::GetU16(m_body + m_nCursor);
	m_nCursor	+= 2;
	return data;
}

I16 Packet::ReadI16()
{
	I16 data = PacketUtil::GetI16(m_body + m_nCursor);
	m_nCursor	+= 2;

	return data;
}

U32 Packet::ReadU32()
{
	U32 data = PacketUtil::GetU32(m_body + m_nCursor);
	m_nCursor	+= 4;
	return data;
}

I32 Packet::ReadI32()
{
	I32 data = PacketUtil::GetI32(m_body + m_nCursor);
	m_nCursor	+= 4;

	return data;
}

U64 Packet::ReadU64()
{
	U64 data = PacketUtil::GetU64(m_body + m_nCursor);
	m_nCursor	+= 8;
	return data;
}

I64 Packet::ReadI64()
{
	I64 data = PacketUtil::GetI64(m_body + m_nCursor);
	m_nCursor	+= 8;

	return data;
}

F32 Packet::ReadF32()
{
	I32 val = ReadI32();
	F32 data = *(F32*)(&val);

	return data;
}

F64 Packet::ReadF64()
{
	I64 val = ReadI64();
	F64 data = *(F64*)(&val);

	return data;
}

String Packet::ReadString()
{
	U16 len = ReadU16();

	I8* data = new I8[len + 1];
	memset(data, '\0', len + 1);
	memcpy(data, m_body + m_nCursor, len);

	m_nCursor	+= len;

	String str(data);
	free(data);

	return str;
}

BYTE* Packet::ReadBytes(U32* len)
{
	*len = GetPacketLen();

	BYTE* buf = new BYTE[*len];
	memcpy(buf, m_body, *len);

	return buf;
}

File: pytools/examples/lua53/client/network/packet.lua
-- local M = {}
-- local moduleName = ...
-- _G[moduleName] = M
-- setmetatable(M, {__index = _G})
-- local _ENV = M
packet = {}
setmetatable(packet, {__index = _G})
local _ENV = packet


function new(buff)
	local tab_obj = {}
	setmetatable(tab_obj, {__index = packet})
	tab_obj:init(buff)
	return tab_obj
end

function init(self, buffers)
	buffers = buffers or ""
	self.buffers		= buffers	-- 包内容字节
	self.pos			= 1			-- 包内容解到哪个位置了
end


function encode(self, packetId)
	self.buffers = string.pack(">H>A", packetId, self.buffers)
	local packeLen = string.len(self.buffers)
	self.buffers = string.pack(">H>A", packeLen, self.buffers)

	return self.buffers
end


function writeU8(self, wu8)
	self.buffers = string.pack(">A>b", self.buffers, wu8)
end
function writeI8(self, wi8)
	self.buffers = string.pack(">A>c", self.buffers, wi8)
end

function writeU16(self, wu16)
	self.buffers = string.pack(">A>H", self.buffers, wu16)
end
function writeI16(self, wi16)
	self.buffers = string.pack(">A>h", self.buffers, wi16)
end

function writeU32(self, wu32)
	self.buffers = string.pack(">A>I", self.buffers, wu32)
end
function writeI32(self, wi32)
	self.buffers = string.pack(">A>i", self.buffers, wi32)
end

function writeU64(self, wu64)
	self.buffers = string.pack(">A>L", self.buffers, wu64)
end
function writeI64(self, wi64)
	self.buffers = string.pack(">A>l", self.buffers, wi64)
end

function writeF32(self, wf32)
	self.buffers = string.pack(">A>f", self.buffers, wf32)
end

function writeF64(self, wf64)
	self.buffers = string.pack(">A>d", self.buffers, wf64)
end

function writeString(self, wstring)
	local stringLen = string.len(wstring)
	self.buffers = string.pack(">A>H>A", self.buffers, stringLen, wstring)
end

function writeBytes(self, bytes)
	self.buffers = string.pack(">A2", self.buffers, bytes)
end

	
function readU8(self)
	self.pos, ru8 = string.unpack(self.buffers, ">b", self.pos)
	return tonumber(ru8)
end
function readI8(self)
	self.pos, ri8 = string.unpack(self.buffers, ">c", self.pos)
	return tonumber(ri8)
end
		
function readU16(self)
	self.pos, ru16 = string.unpack(self.buffers, ">H", self.pos)
	return tonumber(ru16)
end
function readI16(self)
	self.pos, ri16 = string.unpack(self.buffers, ">h", self.pos)
	return tonumber(ri16)
end
		
function readU32(self)
	self.pos, ru32 = string.unpack(self.buffers, ">I", self.pos)
	return tonumber(ru32)
end
function readU32(self)
	self.pos, ri32 = string.unpack(self.buffers, ">i", self.pos)
	return tonumber(ri32)
end

function readU64(self)
	self.pos, ru64 = string.unpack(self.buffers, ">L", self.pos)
	return tonumber(ru64)
end
function readI64(self)
	self.pos, ri64 = string.unpack(self.buffers, ">l", self.pos)
	return tonumber(ri64)
end

function readF32(self)
	self.pos, rF32 = string.unpack(self.buffers, ">f", self.pos)
	return tonumber(rF32)
end

function readF64(self)
	self.pos, rF64 = string.unpack(self.buffers, ">d", self.pos)
	return tonumber(rF64)
end
		
function readString(self)
	stringLen	= self:readU16()
	self.pos, rString = string.unpack(self.buffers, ">A" .. tostring(stringLen), self.pos)
	return tostring(rString)
end

function readBytes(self)
	return self.buffers
end
